numpy/core/src/umath/ufunc_object.c

```diff
@@ -2089,6 +2107,124 @@ PyUFunc_GeneralizedFunction(PyUFuncObject *ufunc,
+            if (ufunc->core_num_dim_ix > 1) {
+                PyErr_SetString(PyExc_ValueError, "axis should be a list");
+                retval = -1;
+                goto fail;
+            }
+            axis_rqst = (int)PyInt_AsSsize_t(axis);
+            if (axis_rqst == -1 && PyErr_Occurred()) {
+                PyErr_Clear();
+                PyErr_SetString(PyExc_ValueError,
+                                "axis should be a list or a single integer");
+                retval = -1;
+                goto fail;
+            }
+        }
+        for (i = 0; i < nop; ++i) {
+            int op_ndim, iax, rqst, dflt;
+            int num_dims = ufunc->core_num_dims[i];
+            if (op[i] != NULL) {
+                op_ndim = PyArray_NDIM(op[i]);
+            }
+            else {
+                op_ndim = broadcast_ndim + ufunc->core_num_dims[i];
+            }
+            /* initialize remap to straight one to one */
+            for (j = 0; j < op_ndim; j++) {
+                remap_axis[i][j] = j;
+            }
+            if (axis_list) {
+                int n;
+                axis_item = PyList_GET_ITEM(axis, i);
+                if (axis_item == Py_None) { /* no remap for this operand */
+                    continue;
+                }
+                axis_tuple = PyTuple_Check(axis_item);
+                if (axis_tuple) {
+                    n = PyTuple_Size(axis_item);
+                }
+                else { /* try as integer */
+                    rqst = (int)PyInt_AsSsize_t(axis_item);
+                    if (rqst == -1 && PyErr_Occurred()) {
+                        PyErr_Clear();
+                        n = -1;  /* to give descriptive error message below */
+                    }
+                    else {
+                        n = 1;
+                    }
+                }
+                if (n != num_dims) {
+                    if (num_dims == 1) {
+                        PyErr_Format(PyExc_ValueError,
+                                     "axis list element %d should be a tuple "
+                                     "with a single elements, an integer, or "
+                                     "None", i);
+                    }
+                    else {
+                        PyErr_Format(PyExc_ValueError,
+                                     "axis list element %d should be a tuple "
+                                     "with %d elements or None", i, num_dims);
+                    }
+                    retval = -1;
+                    goto fail;
+                }
+            } else {
+                rqst = axis_rqst; /* use same axis for all operands */
+            }
+            for (iax = num_dims - 1; iax >= 0; --iax) {
+                if (axis_tuple) {
+                    rqst = (int)PyInt_AsSsize_t(PyTuple_GET_ITEM(axis_item,
+                                                                 iax));
+                    if (rqst == -1 && PyErr_Occurred()) {
```

Comment:
I think there's an `error_converting(rqst)` function in `common.h` that we'd normally use here